_posts/2021-04-15-testing-react-tracking-with-jest-and-enzyme.markdown

```diff
@@ -0,0 +1,242 @@
+
+# The final solution
+
+Our mock ended up looking like this:
+
+```ts
+window.analytics = {
+  track: jest.fn(),
+}
+
+jest.mock("react-tracking", () => ({
+  useTracking: jest.fn(),
+  track: () => (children): typeof children => children,
+}))
+
+const trackEvent = jest.fn((args) => {
+  args.callback && args.callback()
+})
+;(useTracking as jest.Mock).mockImplementation(() => ({
+  trackEvent,
+}))
+```
+
+Let's break that down section by section. First:
+
+```ts
+window.analytics = {
+  track: jest.fn(),
+}
+```
+
+As noted above, all of our tracking calls assume `window.analytics` exists and that it has a `.track()` method. We
+started by mocking that setup.
+
+Next:
+
+```ts
+jest.mock("react-tracking", () => ({
+  useTracking: jest.fn(),
+  track: () => (children): typeof children => children,
+}))
+```
+
+Here we mock the `react-tracking` package and two specific methods it exports, `useTracking` and `track`. We made
+`useTracking` a Jest function - we'll flesh it out further a few lines farther down in the file.
+
+Then there's the mocking of `track`. To put it in words, our mock is: a function that returns a function that takes
+in `children` and returns those `children`. That might sound like gibberish at first blush, but essentially what
+we're doing is mocking the function composition we performed earlier when creating `TrackerContextCohesion`. We
+needed something that was the same shape as `react-tracking`'s `track()`, but we don't care about overriding
+`dispatch` in our mocks.
+
+Last:
+
+```ts
+const trackEvent = jest.fn((args) => {
+  args.callback && args.callback()
+})
+;(useTracking as jest.Mock).mockImplementation(() => ({
+  trackEvent,
+}))
+```
+
+`trackEvent` is a mock function that takes in an `args` object and executes `args.callback()` if it exists. We then
+update our `useTracking` mock to make it return a function that returns an object with a `trackEvent` property.
+What a mouthful! That sounds super confusing, but remember that we're trying to mock something that we actually use
+like this:
+
+```ts
+const { trackEvent } = useTracking()
+```
+
+So basically, our goal was to mock `trackEvent` and we needed to emulate the shape it has when it's exported by
+`react-tracking`. Hopefully that makes things a little clearer.
+
+After some tinkering and eventually getting the mocks to work in a single test file, we moved these mocked
+functions to a `setup.ts` file that all of our Jest tests load automatically. We chose to make these mocks
+available to all tests because then we wouldn't get surprising test failures if we, say, forgot that we were making
+a tracking call in a component and didn't explicitly mock the tracking calls in those tests.
+
+At the end of the day, we can use these mocked calls in our test files by doing the following:
+
+```ts
+import { useTracking } from "react-tracking"
+
+// This only works because we mock tracking in setup.ts
+const { trackEvent } = useTracking()
```

Comment:
This is a bit strange because it breaks the [rules of hooks](https://reactjs.org/docs/hooks-rules.html)

Reply:
That made me curious. This exists https://github.com/testing-library/react-hooks-testing-library#usecountertestjs, but this feels like this is testing the hook, not checking expectations on a hook. I found https://chanonroy.medium.com/mocking-hooks-for-testing-with-jest-and-react-testing-library-d34505616d12 this one. Not sure if that's better.

Reply:
Hmmm yeah...I kinda think it's fine for testing purposes?

Reply:
> I kinda think it's fine for testing purposes

Rules of hooks can never be broken unfortunately! It's kinda a weird flag for those who look at react code all day.